### PR TITLE
Fix incorrect problem length when reporting compilation problems

### LINKED ISSUE #28554: Incorrect java warning/error location in new Problems API
### Current Behavior

Take something like...
```java
public class WarningsCreator {

  private static int idx = 0;

  private void createIndirectAccessWarning() {
    int idx = new WarningsCreator().idx;
  }
}
```

with compile settings...
```groovy
tasks.withType(JavaCompile).configureEach {
	options.compilerArgs << '-Xlint:all'
}
```

Java will throw a warning: `static variable should be qualified by type name, WarningsCreator, instead of by an expression`

The `javax.tools.Diagnostic` class will have the following position data for this warning...

```java
    int idx = new WarningsCreator().idx;
              ^                               startPosition (in file)
                                      ^       endPosition (in file)
                                   ^          position (in file)
                                   ^          columnNumber (in line)
```
This is because Java sets `startPosition` and `endPosition` to the whole `new WarningsCreator().idx` range but it sets `columnNumber`, `position` to point to the start of `.idx;`

The `offsetInFileLocation` is being reported using `startPosition` and `endPosition - startPosition` (so the length is 25).  This is correct.

But the **`lineInFileLocation`** is being reported incorrectly because the `length` is calculated using `endPosition - startPosition` (i.e. 25) when it **should be using `endPosition - position`** (i.e. 4)

### Expected Behavior

The length in `lineInFileLocation` should be correct

### Context (optional)

Trying to highlight the errors/warning correctly in an IDE

### Steps to Reproduce

Use the simple java snippet above with the `xlint` settings provided.

### Gradle version

8.8-20240323001552+0000

### Build scan URL (optional)

_No response_

### Your Environment (optional)

Windows 11

### PR DESCRIPTION
**Pull request by hegyibalint**
Fixes #28554

### CODE REVIEW

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 1]
```diff
@@ -124,9 +125,10 @@ private static void addLocations(ProblemSpec spec, Diagnostic<? extends JavaFile
 
             // If we know the offsets ...
             // (offset doesn't require line and column to be set, hence the separate check)
-            if (0 < start && start <= end) {
+            // (documentation says that getEndPosition() will be NOPOS iff getPosition() is NOPOS)
+            if (0 < position) {
                 // ... we can report the start and extent
-                spec.offsetInFileLocation(resourceName, start, end - start);
+                spec.offsetInFileLocation(resourceName, start, end - position);
```
Comment (Arthurm1) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:131]: Not sure if you made this PR before my comment but if you're going to change this to `end - position` then `start` should also be changed to `position`

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java [hunk 2]
```diff
@@ -97,21 +97,22 @@ private static void addLocations(ProblemSpec spec, Diagnostic<? extends JavaFile
...
             // If we know the line ...
-            if (line > 0) {
+            if (0 < line) {
                 // ... and the column ...
-                if (column > 0) {
+                if (0 < column) {
                     // ... and we know how long the error is (i.e. end - start)
-                    // (end should be greater than start, so we can prevent negative lengths)
-                    if (0 < start && start <= end) {
+                    // (documentation says that getEndPosition() will be NOPOS iff getPosition() is NOPOS)
```
Comment (Arthurm1) [inline platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:112]: iff -> if

Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java:112]: Maybe I went to mathematical here, official doc says "if and only if".

File: platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java (hunk 1 above)
Comment (hegyibalint) [reply platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/DiagnosticToProblemListener.java]: Absolutely right, fixed!

**Review by donat — APPROVED** [review]

**Pull request merged**